File: src/tracks/sibling/status.ts
import * as crypto from "node:crypto";
import * as fs from "node:fs";
import * as path from "node:path";
import { defaultManifestPath, readManifest } from "../../manifest";
import { resolveRepoRoot } from "../../repo-root";
import { siblingConfigPath, type SiblingConfig } from "./init";

/**
 * `omc-sync status --track sibling`
 *
 * Per-file sync report for the sibling track: for each manifest-listed
 * path, compares `.omc/<path>` (the anchor repo's copy) against
 * `<clonePath>/<path>` (the sibling clone's copy) by content hash. Mirrors
 * the shadow track's per-file report (`../shadow/status.ts`), classifying
 * each path as one of:
 *   - "in sync": present in both, with matching content.
 *   - "pending (local changes)": present in both, but content differs.
 *   - "pending (never synced)": present locally, absent from the clone.
 *   - "missing locally": present in the clone, absent locally.
 */

function readSiblingConfig(repoRoot: string): SiblingConfig {
  const configPath = siblingConfigPath(repoRoot);
  if (!fs.existsSync(configPath)) {
    throw new Error(
      `status --track sibling: no sibling config found at ${configPath} — run \`omc-sync init --track sibling\` first`,
    );
  }

  const raw = JSON.parse(fs.readFileSync(configPath, "utf8")) as {
    sibling?: SiblingConfig;
  };
  if (!raw.sibling) {
    throw new Error(
      `status --track sibling: ${configPath} has no "sibling" entry — run \`omc-sync init --track sibling\` first`,
    );
  }

  return raw.sibling;
}

function sha256(content: Buffer): string {
  return crypto.createHash("sha256").update(content).digest("hex");
}

export function run(_args: string[]): void {
  const repoRoot = resolveRepoRoot();
  const { clonePath } = readSiblingConfig(repoRoot);
  const manifestPaths = readManifest(defaultManifestPath(repoRoot));

  let inSync = 0;
  let pendingLocal = 0;
  let pendingNeverSynced = 0;
  let missingLocally = 0;

  for (const relPath of manifestPaths) {
    const localPath = path.join(repoRoot, ".omc", relPath);
    const clonedPath = path.join(clonePath, relPath);
    const localExists = fs.existsSync(localPath);
    const clonedExists = fs.existsSync(clonedPath);

    let state: string;
    if (localExists && clonedExists) {
      const localHash = sha256(fs.readFileSync(localPath));
      const clonedHash = sha256(fs.readFileSync(clonedPath));
      if (localHash === clonedHash) {
        state = "in sync";
        inSync++;
      } else {
        state = "pending (local changes)";
        pendingLocal++;
      }
    } else if (localExists) {
      state = "pending (never synced)";
      pendingNeverSynced++;
    } else if (clonedExists) {
      state = "missing locally";
      missingLocally++;
    } else {
      // Neither locally present nor ever synced — closest fit of the four
      // reported states is "never synced".
      state = "pending (never synced)";
      pendingNeverSynced++;
    }

    process.stdout.write(`omc-sync: ${relPath}: ${state}\n`);
  }

  process.stdout.write(
    `omc-sync: ${manifestPaths.length} file(s) tracked — ${inSync} in sync, ${pendingLocal} pending (local changes), ${pendingNeverSynced} pending (never synced), ${missingLocally} missing locally\n`,
  );
}

File: go/internal/tracks/sibling/status.go
package sibling

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"path/filepath"

	"plan-sync/go/internal/manifest"
)

// Status is the Go port of src/tracks/sibling/status.ts's run():
// `plan-sync status --track sibling [--root <dir>]`.
//
// Per-file sync report: for each manifest-listed path, the anchor repo's
// copy is compared against the sibling clone's copy by content hash and
// classified as one of
//
//	"in sync"                 — present in both, matching content.
//	"pending (local changes)" — present in both, content differs.
//	"pending (never synced)"  — present locally, absent from the clone.
//	"missing locally"         — present in the clone, absent locally.
//
// A path present in neither falls back to "pending (never synced)", the
// closest fit of the four reported states.
func Status(argv []string) error {
	repoRoot, rootDir, err := resolveContext(argv)
	if err != nil {
		return err
	}

	cfg, err := readSiblingConfig("status", repoRoot, rootDir)
	if err != nil {
		return err
	}
	clonePath := cfg.ClonePath

	// Resolved (live pattern re-evaluation against the current filesystem,
	// plus every literal entry even when currently absent), not the raw
	// manifest lines — this is "what should be reported right now".
	manifestPaths := manifest.ResolveManifestSyncCandidates(manifestPathFor(repoRoot, rootDir))

	inSync := 0
	pendingLocal := 0
	pendingNeverSynced := 0
	missingLocally := 0

	for _, relPath := range manifestPaths {
		localPath := filepath.Join(repoRoot, rootDir, relPath)
		clonedPath := filepath.Join(clonePath, relPath)
		localExists := fileExists(localPath)
		clonedExists := fileExists(clonedPath)

		var state string
		switch {
		case localExists && clonedExists:
			localHash, localErr := sha256File(localPath)
			clonedHash, clonedErr := sha256File(clonedPath)
			if localErr == nil && clonedErr == nil && localHash == clonedHash {
				state = "in sync"
				inSync++
			} else {
				state = "pending (local changes)"
				pendingLocal++
			}
		case localExists:
			state = "pending (never synced)"
			pendingNeverSynced++
		case clonedExists:
			state = "missing locally"
			missingLocally++
		default:
			state = "pending (never synced)"
			pendingNeverSynced++
		}

		fmt.Fprintf(os.Stdout, "plan-sync: %s: %s\n", relPath, state)
	}

	fmt.Fprintf(os.Stdout,
		"plan-sync: %d file(s) tracked — %d in sync, %d pending (local changes), %d pending (never synced), %d missing locally\n",
		len(manifestPaths), inSync, pendingLocal, pendingNeverSynced, missingLocally)
	return nil
}

// sha256File hashes a file's contents. An unreadable file is reported as an
// error, which Status treats as "not identical" rather than aborting the
// whole report.
func sha256File(path string) (string, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256(content)
	return hex.EncodeToString(sum[:]), nil
}
